src/goto-instrument/contracts/dynamic-frames/dfcc.h

```diff
@@ -0,0 +1,235 @@
+void dfcc(
+  goto_modelt &goto_model,
+  const std::string &harness_id,
+  const std::map<std::string, std::string> &to_check,
+  const bool allow_recursive_calls,
+  const std::map<std::string, std::string> &to_replace,
+  const bool apply_loop_contracts,
+  const std::set<std::string> &exclude_from_nondet_static,
+  messaget &log);
+
+/// \ingroup dfcc-module
+/// \brief Entry point into the contracts transformation
+class dfcct
+{
+public:
+  /// Class constructor
+  /// \param goto_model GOTO model to transform
+  /// \param log logger to use for debug/warning/error messages
+  dfcct(goto_modelt &goto_model, messaget &log);
+
+  /// Applies function contracts and loop contracts transformation to GOTO model
+  /// using the dynamic frame condition checking approach.
+  ///
+  /// Functions to check/replace are explicitly mapped to contracts.
+  /// When checking function `foo` against contract `bar`, we require the
+  /// actual contract symbol to exist as `contract::bar` in the symbol table.
+  ///
+  /// \param harness_id proof harness name, must be the entry point of the model
+  /// \param to_check functions-to-contract mapping for contract checking
+  /// (must contain at most one element)
+  /// \param allow_recursive_calls Allow the checked function to be recursive
+  /// \param to_replace functions-to-contract mapping for replacement
+  /// \param apply_loop_contracts apply loop contract transformations iff true
+  /// \param to_exclude_from_nondet_static set of symbols to exclude when
+  /// havocing static program symbols.
+  ///
+  /// Transformation steps:
+  /// - check preconditions on existence and absence of clash between harness,
+  /// functions and contract symbols parameters.
+  /// - link the goto model to the standard library
+  /// - instrument the harness function.
+  /// - partition function symbols of the model into
+  ///   - `contract::` symbols
+  ///   - assigns/frees clause specification functions (not instrumented)
+  ///   - all other functions (instrumented)
+  /// - swap-and-wrap/instrument functions to check with their contract
+  /// - swap-and-wrap/instrument functions to replace with their contract
+  /// - swap-and-wrap all discovered function pointer contracts with themselves
+  ///   (replacement mode)
+  /// - instrument all remaining functions GOTO model
+  /// - (this includes standard library functions).
+  /// - specialise the instrumentation functions by unwiding loops they contain
+  /// to the maximum size of any assigns clause involved in the model.
+  void transform_goto_model(
+    const std::string &harness_id,
+    const std::map<std::string, std::string> &to_check,
+    const bool allow_recursive_calls,
+    const std::map<std::string, std::string> &to_replace,
+    const bool apply_loop_contracts,
+    const std::set<std::string> &to_exclude_from_nondet_static);
+
+protected:
+  goto_modelt &goto_model;
+  messaget &log;
+  message_handlert &message_handler;
```

Comment:
I used the same interface as `code_contractst` which receives a `messaget` (from `goto_instrument_parse_optionst`), but since some other classses/functions require a `message_handlert` I just unpack it in advance. 

Looking at the rest of the codebase, sometimes a `messaget` gets passed around, sometimes it is a `message_handlert`, in which case a new `messaget(handler)` needs to be built anyway.

What would be the recommended pattern ?

Reply:
It should always be a `message_handlert`, unless you really want to carry over state of the output stream (note that a `messaget` has an `ostringstream`-derived member).

Reply:
fixed